メッセージ機能の非同期通信

yabecchi312 opened the pull request:
# what
メッセージ機能の非同期通信化

# why
ux向上のため

Commit:
first commit

Commit:
debug new message

フォームが送信された時に、イベントが発火しているかどうかの確認

Commit:
add ajax

イベントが発火したときにAjaxを使用して、messages#createが動くようにする

Commit:
describe create action respond_to

コメントを保存し、respond_toを使用してHTMLとJSONの場合で処理を分けるためにmessages_controllerのcreate
アクションにrepond_toを記述

Commit:
add create.json.jbuilder

jbuilderを使用して、作成したメッセージをJSON形式で返す

Commit:
add done and fail method message.js

・返ってきたJSONをdoneメソッドで受取り、HTMLを作成する。
・作成したHTMLをメッセージ画面の一番下に追加する
・HTMLを追加した分、メッセージ画面を下にスクロールする
・ 非同期に失敗した場合の処理も準備する

MrUrakou commented on app/views/messages/_message.html.haml:
```diff
@@ -3,7 +3,7 @@
       = message.user.name
     .upper-message__date
-      = message.created_at
+      = message.created_at.strftime('%Y/%m/%d %H:%M')
```
timeformat railsでググって日付の表示の仕方を工夫してください。

yabecchi312 replied:
timeformatに修正しました！

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,47 @@
+      var html = buildHTML(data);
+      $('.messages').append(html)
+      $('.form').val('')
+      $('.messages').animate({scrollTop: $('.messages')[0].scrollHeight},'fast' )
```
スクロール機能は関数化して呼び出しましょう。

yabecchi312 replied:
関数化しました！

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,47 @@
+    .done(function(data){
+      var html = buildHTML(data);
+      $('.messages').append(html)
+      $('.form').val('')
```
resetメソッド使いましょう。

yabecchi312 replied:
resetメソッドを加えました。

Commit:
alter time format and add reset method

・新規でinitiarizeのファイルの中にtime_formatsを作成して、create_atのあとを.to_sを使って時刻を生成。
・resetメソッドを.doneの最後に記述
・scrollを関数化して最初に定義

MrUrakou commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,48 @@
+    var formData = new FormData(this);
+    var url = $(this).attr('action')
+    function scroll(messages){
+      messages.animate({scrollTop: $('.messages')[0].scrollHeight},'fast')
```
$('.messages')に対して直接animateメソッドかければ引数不要ですね。
あとここに関数の記述するのではなく、buildHTMLの関数みたいに外に切り出しましょう。

yabecchi312 replied:
わかりました。
修正します。

Commit:
alter scroll

function scrollを外に出して、引数を消しました。

MrUrakou commented on app/views/messages/create.json.jbuilder:
```diff
@@ -0,0 +1,5 @@
+json.content  @message.content
+json.image  @message.image.url
+json.user_id  @message.user.id
```
jsで使用しないものは削除しましょう。

yabecchi312 replied:
わかりました。
user_idは使わないので削除しました。

Commit:
delete user_id create.json.jbuilder

user_idはjsで使用しないので削除
